zkevm-circuits/src/root_circuit/aggregation.rs

```diff
@@ -0,0 +1,698 @@
+/// KZG deciding key
+pub type KzgDk<M> = KzgDecidingKey<M>;
+/// Plonk succinct verifier with `KzgAs`
+pub type PlonkSuccinctVerifier<M> = verifier::plonk::PlonkSuccinctVerifier<KzgAs<M>>;
+/// Plonk verifier with `KzgAs` and `LimbsEncoding<LIMBS, BITS>`.
+pub type PlonkVerifier<M> = verifier::plonk::PlonkVerifier<KzgAs<M>, LimbsEncoding<LIMBS, BITS>>;
+
+const T: usize = 5;
+const RATE: usize = 4;
+const R_F: usize = 8;
+const R_P: usize = 60;
+
+/// `BaseFieldEccChip` with hardcoded `LIMBS` and `BITS` serving as `EccChip`
+/// for `Halo2Loader`.
+pub type EccChip<C> = halo2_wrong_ecc::BaseFieldEccChip<C, LIMBS, BITS>;
+/// `Halo2Loader` with hardcoded `EccChip`.
+pub type Halo2Loader<'a, C> = loader::halo2::Halo2Loader<'a, C, EccChip<C>>;
+/// `PoseidonTranscript` with hardcoded parameter with 128-bits security.
+pub type PoseidonTranscript<C, S> =
+    transcript::halo2::PoseidonTranscript<C, NativeLoader, S, T, RATE, R_F, R_P>;
+
+/// Snark contains the minimal information for verification
+#[derive(Clone, Copy)]
+pub struct Snark<'a, C: CurveAffine> {
+    protocol: &'a PlonkProtocol<C>,
+    instances: &'a Vec<Vec<C::Scalar>>,
+    proof: &'a [u8],
+}
+
+impl<'a, C: CurveAffine> Snark<'a, C> {
+    /// Construct `Snark` with each field.
+    pub fn new(
+        protocol: &'a PlonkProtocol<C>,
+        instances: &'a Vec<Vec<C::Scalar>>,
+        proof: &'a [u8],
+    ) -> Self {
+        Self {
+            protocol,
+            instances,
+            proof,
+        }
+    }
+}
+
+impl<'a, C: CurveAffine> From<Snark<'a, C>> for SnarkWitness<'a, C> {
+    fn from(snark: Snark<'a, C>) -> Self {
+        Self {
+            protocol: snark.protocol,
+            instances: Value::known(snark.instances),
+            proof: Value::known(snark.proof),
+        }
+    }
+}
+
+/// SnarkWitness
+#[derive(Clone, Copy)]
+pub struct SnarkWitness<'a, C: CurveAffine> {
+    protocol: &'a PlonkProtocol<C>,
+    instances: Value<&'a Vec<Vec<C::Scalar>>>,
+    proof: Value<&'a [u8]>,
+}
+
+impl<'a, C: CurveAffine> SnarkWitness<'a, C> {
+    /// Construct `SnarkWitness` with each field.
+    pub fn new(
+        protocol: &'a PlonkProtocol<C>,
+        instances: Value<&'a Vec<Vec<C::Scalar>>>,
+        proof: Value<&'a [u8]>,
+    ) -> Self {
+        Self {
+            protocol,
+            instances,
+            proof,
+        }
+    }
+
+    /// Returns `SnarkWitness` with all witness as `Value::unknown()`.
+    pub fn without_witnesses(&self) -> Self {
+        SnarkWitness {
+            protocol: self.protocol,
+            instances: Value::unknown(),
+            proof: Value::unknown(),
+        }
+    }
+
+    /// Returns protocol as reference.
+    pub fn protocol(&self) -> &PlonkProtocol<C> {
+        self.protocol
+    }
+
+    /// Returns proof as reference.
+    pub fn proof(&self) -> Value<&[u8]> {
+        self.proof
+    }
+
+    fn loaded_instances<'b>(
+        &self,
+        loader: &Rc<Halo2Loader<'b, C>>,
```

Comment:
Maybe I'll create an issue on https://github.com/privacy-scaling-explorations/snark-verifier then? My point is that I don't really see why you need to have RC in general. 

But of course you have the full picture. So that's why I asked! :)

Reply:
Sure! it’d be nice to get rid of the `Rc`, I remember it’s due to lack of GAT so I need to use `Rc`, but perhaps you’d know how to work around that.